### PR TITLE
avm1: Correct and cleanup `Array.sort` and `Array.sortOn`

### LINKED ISSUE #4623: Onda Educa: Tile randomization not working
**Describe the bug**
The tiles of the memory matching games are supposed to randomize positions, but in Ruffle they don't.
This of course allows students to memorize the puzzles, which is counterproductive. 

**Expected behavior**
Tiles should be able to randomize positions.

Web builds: Extension and Self-Host

**What platform are you using?**
 Self-Host with Chromium and Mozilla based browsers.

Tagging #1182 for reference, but I think this problem is a little more nuanced. These games are rounding the result of Math.random(), and that is being used in an anonymous callback function to array.sort(). This would indicate to me that at one point in time, floating point numbers greater than one were expected from Math.random(), not values between 0 and 1. Furthermore, since the anon function has no arguments, looking at Ruffle's code, I would assume it's just going to stick with default sorting behavior, which wouldn't accomplish anything useful, nor would setting the sort behavior to any other mode even if the code does respect the 0 or 1 as a bitflag.

[bingo.zip](https://github.com/ruffle-rs/ruffle/files/6681528/bingo.zip)

### PR DESCRIPTION
**Pull request by relrelb**
* Mimic Flash's quicksort algorithm, rather than using Rust's
`sort_unstable_by`.
* Rename `flags` to `options`, as they are named in AS2 reference.
* Organize different sorting options using 3 simple functions: `sort_compare`, `sort_compare_custom` and `sort_on_compare`.

Fixes #4623, and maybe others.

### CODE REVIEW

**Comment by relrelb** [comment]
Converted to draft as I discovered some issues with the current code.

File: core/src/avm1/globals/array.rs [hunk 1]
```diff
@@ -417,273 +425,331 @@ fn sort<'gc>(
...
     };
-    let flags = SortFlags::from_bits_truncate(flags);
 
-    let string_compare_fn = if flags.contains(SortFlags::CASE_INSENSITIVE) {
-        sort_compare_string_ignore_case
+    let compare_fn = if let Some(compare_fn) = compare_fn {
+        sort_compare_custom(compare_fn)
     } else {
-        sort_compare_string
+        Box::new(sort_comapre)
```
Comment (Herschel) [inline core/src/avm1/globals/array.rs:448]: typo: `sort_comapre` -> `sort_compare`

Comment (relrelb) [reply core/src/avm1/globals/array.rs]: Oops. Nice catch.

File: core/src/avm1/globals/array.rs [hunk 2]
```diff
@@ -417,273 +425,331 @@ fn sort<'gc>(
...
+        Ok(result.partial_cmp(&0.0).unwrap_or(DEFAULT_ORDERING))
+    })
+}
 
-            if flags.contains(SortFlags::NUMERIC) {
-                sort_compare_numeric(flags.contains(SortFlags::CASE_INSENSITIVE))
-            } else {
-                Box::new(string_compare_fn)
+/// Create a compare function based on field names and options.
+fn sort_on_comapre<'a, 'gc>(fields: &'a [(AvmString<'gc>, SortOptions)]) -> CompareFn<'a, 'gc> {
```
Comment (Herschel) [inline core/src/avm1/globals/array.rs:581]: Typo

Comment (relrelb) [reply core/src/avm1/globals/array.rs]: Oops again. Thanks!

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**